compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -204,11 +206,58 @@ fn write_path(out: &mut String, path: &[PathElem]) {
             // not the root.
             Deref => write!(out, ".<deref>"),
             DynDowncast => write!(out, ".<dyn-downcast>"),
+            Vtable => write!(out, ".<vtable>"),
         }
         .unwrap()
     }
 }
 
+/// Represents a set of `Size` values as a sorted list of ranges.
+// These are (offset, length) pairs, and they are sorted and mutually disjoint,
+// and never adjacent (i.e. there's always a gap between two of them).
+#[derive(Debug, Clone)]
+pub struct RangeSet(Vec<(Size, Size)>);
```

Comment:
This is kind of like `IntervalSet` in `rustc_index`, except we need u64 and not just u32.  (Also `IntervalSet` has a `domain` which we don't need, but we could just set `u64::MAX`.)

Alternatively, https://docs.rs/rangemap/1.5.1/rangemap/set/struct.RangeSet.html would probably also do it. But that would add a new dependency.

Reply:
Maybe I should try making `IntervalSet` generic over a `T: Into<u64>`? That will make it less efficient (computation-wise, not storage-wise) on a 32bit system. Not sure if we care.

The alternative to `Into<u64>` is to basically require the equivalent of a `Num` trait, which might get awkward, but I haven't tried.

Reply:
Ah no `Into<u64>` is not enough. We need at least things like `max`, though that is easy via `PartialOrd`. More problematic is the fact that IntervalSet uses inclusive ranges for some reason, so it has to do a lot of annoying `+1`. My implementation uses right-exclusive ranges which, as usual, makes the arithmetic a lot nicer -- but maybe someone benchmarked this and  anyway I'm not going to re-do IntervalSet now.^^

Cc @Mark-Simulacrum since you seem to have added `IntervalSet` a couple of years ago.

Reply:
FWIW I'm perfectly happy to see new data structures land in random places initially with highly specific implementations. If someone finds a use for it somewhere else, that's the point at which it becomes worth it to make things reusable.